Add JSON processor documentation (#5982)
opensearch-project/documentation-website: _ingest-pipelines/processors/json.md

---
layout: default
title: JSON
parent: Ingest processors
nav_order: 170
---

# JSON processor

> The `json` processor serializes a string-valued field into a map of maps, which can be useful for various data processing and enrichment tasks.

Reviewer: Is "string-valued" the right phrasing here, or should it just be "string value field"?

> The `json` processor serializes a string-valued field into a map of maps, which can be useful for various data processing and enrichment tasks.

Author: ```suggestion
The `json` processor serializes a string value field into a map of maps, which can be useful for various data processing and enrichment tasks.
```

The following is the syntax for the `json` processor:

```json
{
  "processor": {
    "json": {
      "field": "<field_name>",
      "target_field": "<target_field_name>",
      "add_to_root": <boolean>
    }
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `json` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`field` | Required | The name of the field containing the JSON-formatted string to be deserialized.

> `target_field` | Optional | The name of the field where the deserialized JSON data will be stored. If `target_field` is unspecified, the value is stored in field. If `target_field` exists, it is overwritten.

Reviewer: ```suggestion
`target_field` | Optional | The name of the field in which the deserialized JSON data will be stored. If `target_field` is unspecified, then the value is stored in field. If `target_field` exists, it is overwritten.
```

> `target_field` | Optional | The name of the field where the deserialized JSON data will be stored. If `target_field` is unspecified, the value is stored in field. If `target_field` exists, it is overwritten.

Reviewer: The last two sentences here have me a bit confused. Can we check for clarity/accuracy?

> `target_field` | Optional | The name of the field in which the deserialized JSON data will be stored. If `target_field` is unspecified, then the value is stored in field. If `target_field` exists, it is overwritten.

Author: ```suggestion
`target_field` | Optional | The name of the field in which the deserialized JSON data is stored. When not provided, the data is stored in the `field` field. If `target_field` exists, its existing value is overwritten with the new JSON data.
```

Author: @natebower Please see the rewrites for line 10 and 34. Thank you :)

> `add_to_root` | Optional | A boolean flag that determines whether the deserialized JSON data should be added to the root of the document (`true`) or stored in the target_field (`false`). If `add_to_root` is `true`, then `target-field` is invalid. Default value is `false`. 

Reviewer: ```suggestion
`add_to_root` | Optional | A Boolean flag that determines whether the deserialized JSON data should be added to the root of the document (`true`) or stored in the target_field (`false`). If `add_to_root` is `true`, then `target-field` is invalid. Default value is `false`. 
```

`description` | Optional | A description of the processor's purpose or configuration.
`if` | Optional | Specifies to conditionally execute the processor.

> `ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer: ```suggestion
`ignore_failure` | Optional | Specifies to ignore processor failures. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

        "_source": {
          "processed_timestamp": "2024-05-30T15:24:48.064472090Z",
          "raw_data": """{"name":"John","age":30,"city":"New York"}""",
          "parsed_data": {
            "name": "John",
            "city": "New York",
            "age": 30
          }
        },
        "_ingest": {
          "timestamp": "2024-05-30T15:24:48.06447209Z"
        }
      }
    },
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "processed_timestamp": "2024-05-30T15:24:48.064543006Z",
          "raw_data": """{"name":"Jane","age":25,"city":"Los Angeles"}""",
          "parsed_data": {
            "name": "Jane",
            "city": "Los Angeles",
            "age": 25
          }
        },
        "_ingest": {
          "timestamp": "2024-05-30T15:24:48.064543006Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Step 3: Ingest a document 

The following query ingests a document into an index named `my-index`:

```json
POST my-index/_doc?pipeline=my-json-pipeline
{
  "raw_data": "{\"name\":\"John\",\"age\":30,\"city\":\"New York\"}"
}
```
{% include copy-curl.html %}

#### Response

> This response confirms that the document with the JSON data from the `raw_data` field was successfully indexed:

Reviewer: ```suggestion
The response confirms that the document containing the JSON data from the `raw_data` field was successfully indexed:
```

---
layout: default
title: JSON
parent: Ingest processors
nav_order: 170
---

# JSON processor

The `json` processor serializes a string-valued field into a map of maps, which can be useful for various data processing and enrichment tasks.

The following is the syntax for the `json` processor:

```json
{
  "processor": {
    "json": {
      "field": "<field_name>",
      "target_field": "<target_field_name>",
      "add_to_root": <boolean>
    }
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `json` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`field` | Required | The name of the field containing the JSON-formatted string to be deserialized.

> `target_field` | Optional | The name of the field where the deserialized JSON data will be stored. If this parameter is not provided, the deserialized data is added to the root of the document.

Reviewer 2: For the `target_field`, if it's not specified, the value will be stored in the `field` field, and if the specified target_field exists, it will be overwritten.

> `add_to_root` | Optional | A boolean flag that determines whether the deserialized JSON data should be added to the root of the document (`true`) or stored in the target_field (`false`). Default value is `false`. 

Reviewer 2: `add_to_root` and `target_field` are partly exclusive, if `add_to_root` is `true`, `target_field` cannot be set.

`description` | Optional | A description of the processor's purpose or configuration.
`if` | Optional | Specifies to conditionally execute the processor.
`ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`on_failure`| Optional | Specifies a list of processors to run if the processor fails during execution. These processors are executed in the order they are specified. 
`tag` | Optional | An identifier tag for the processor. Useful for debugging in order to distinguish between processors of the same type.

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline named `my-json-pipeline` that uses the `json` processor to process JSON data and enrich the documents with additional information: 

```json
PUT _ingest/pipeline/my-json-pipeline
{
  "description": "Example pipeline using the JsonProcessor",
  "processors": [
    {
      "json": {
        "field": "raw_data",
        "target_field": "parsed_data",

>         "add_to_root": false,

Reviewer 2: `add_to_root` is not needed here because it defaults to `false`.
